contracts/TokenRequest.sol

```diff
@@ -1,161 +1,157 @@
+    bytes32 constant public SET_VAULT_ROLE = keccak256("SET_VAULT_ROLE");
+    bytes32 constant public FINALISE_TOKEN_REQUEST_ROLE = keccak256("FINALISE_TOKEN_REQUEST_ROLE");
+
+    string private constant ERROR_NO_AMOUNT = "TOKEN_REQUEST_NO_AMOUNT";
+    string private constant ERROR_NO_DEPOSIT = "TOKEN_REQUEST_NO_DEPOSIT";
+    string private constant ERROR_ETH_VALUE_MISMATCH = "TOKEN_REQUEST_ETH_VALUE_MISMATCH";
+    string private constant ERROR_TOKEN_TRANSFER_REVERTED = "TOKEN_REQUEST_TOKEN_TRANSFER_REVERTED";
+
+    struct TokenRequest {
+        address requesterAddress;
+        address depositToken;
         uint256 depositAmount;
-        address token;
-        uint256 mintAmount;
-        bool active;
+        uint256 requestAmount;
     }
 
-    Vault public vault;
     TokenManager public tokenManager;
-    Voting public voting;
-    MiniMeToken private token;              //temporary workaround, to show amount of tokens on radspecs's redeem function
-
-    // In case of refund how can we know which pending refund to do?
-    mapping (address => mapping (uint256 => Request)) internal requests;
-    mapping (address => uint256) public requestsLengths;
-
-    event Request(address indexed receiver, address depositToken, uint256 depositAmount, uint256 mintAmount, uint256 voteId, uint256 requestId);
-    event ApprovedRequest(
-        address indexed receiver,
-        address depositToken,
-        uint256 depositAmount,
-        uint256 mintAmount,
-        uint256 voteId,
-        uint256 requestId
-        );
-
-     /**
-    * @notice Initialize TokenRequest app contract
-    * @param _vault Address of the vault
-    * @param _tokenManager TokenManager address
-    * @param _voting Address of the voting
-    */
-    function initialize(Vault _vault, TokenManager _tokenManager, Voting _voting) external onlyInit {
+    address public vault;
+
+    uint256 public nextTokenRequestId;
+    mapping(uint256 => TokenRequest) public tokenRequests; // ID => TokenRequest
+    mapping(address => uint256[]) public addressesTokenRequestIds; // Sender address => List of ID's
+
+    event TokenRequestCreated(uint256 requestId);
+    event TokenRequestRefunded(address refundToAddress, address refundToken, uint256 refundAmount);
+    event TokenRequestFinalised(address requester, address depositToken, uint256 depositAmount, uint256 requestAmount);
+
+    function initialize(address _tokenManager, address _vault) public onlyInit {
         initialized();
 
-        require(isContract(_vault), ERROR_VAULT_NOT_CONTRACT);
-        require(isContract(_tokenManager), ERROR_TOKEN_MANAGER_NOT_CONTRACT);
+        tokenManager = TokenManager(_tokenManager);
+        vault = _vault;
+    }
+
+    /**
+    * @notice Set the Token Manager to `_tokenManager`.
+    * @param _tokenManager The new token manager address
+    */
+    function setTokenManager(address _tokenManager) external auth(SET_TOKEN_MANAGER_ROLE) {
```

Comment:
Do what the same?

Reply:
I mean having functions to set both TokenManager and Vault.

Reply:
I'm not sure, I think there was a requirement that those addresses cannot be changed but I may be incorrect. @lkngtn perhaps can clarify? Also could you clarify if we should have setters for these contracts in the token-request-app as above?

Reply:
Generally I'm of the opinion that initialization parameters should have function to change them, unless changing them would cause specific problems (as opposed to just being potentially risky if they were changed without thinking through the change). A good example of this would be changing the vault associated with a finance app, because changing this would make the history of transactions not make sense (or require some weird investigation). A bad example is that the existing voting app only allows you to change the duration on initialization (for the reason that changing the duration could have security implications if it was shortened after the fact). But in cases where you want to ensure the that a function is never called you can 1) discard the permission, and/or 2) discard the permission and pin the app so it can never be upgraded. 

I don't see any reason why we wouldn't want to allow either the Token Manager or the Payment token to be changed here (though it would probably be very uncommon for these functions to be used). 

I think the same reasoning applies for Redemptions.